### PR TITLE
[ONNX] index_select

### PR DESCRIPTION
**Pull request by cynthia-liu**

### CODE REVIEW

File: pyzoo/zoo/pipeline/api/onnx/mapper/gather.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    def _to_tensor(self):
+        dim = int(self.onnx_attr['axis'])
+
+        assert dim >= 0, "Currently dim>=0 is required."
+        assert dim == 1, "Currently dim=0 is not supported"
+
+        data = self.model_inputs[0].zvalue
+
+        return data.slice(dim=dim, start_index=1, length=1)
```
Comment (zhichao-li) [inline pyzoo/zoo/pipeline/api/onnx/mapper/gather.py]: you need to extract the dim and index from `input` instead of hard code the value.

File: pyzoo/zoo/pipeline/api/onnx/mapper/gather.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+class GatherMapper(OperatorMapper):
+    def __init__(self, node, initializer, _all_tensors):
+        super(GatherMapper, self).__init__(node, initializer, _all_tensors)
+
+    def _to_tensor(self):
+        dim = int(self.onnx_attr['axis'])
+
+        assert dim >= 0, "Currently dim>=0 is required."
+        assert dim == 1, "Currently dim=0 is not supported"
```
Comment (zhichao-li) [inline pyzoo/zoo/pipeline/api/onnx/mapper/gather.py]: remove this?

File: pyzoo/zoo/pipeline/api/onnx/mapper/gather.py (hunk 1 above)
Comment (zhichao-li) [reply pyzoo/zoo/pipeline/api/onnx/mapper/gather.py]: Can we use `index_select` instead of `slice` here?

File: pyzoo/zoo/pipeline/api/onnx/mapper/gather.py (hunk 2 above)
Comment (cynthia-liu) [reply pyzoo/zoo/pipeline/api/onnx/mapper/gather.py]: The parameter 'axis' is in onnx 'gather', I should ignore it?

**Comment by zhichao-li** [comment]
@Deliadd  Please review and help xinqi to refactor this PR.

File: pyzoo/zoo/pipeline/api/onnx/mapper/gather.py (hunk 1 above)
Comment (cynthia-liu) [reply pyzoo/zoo/pipeline/api/onnx/mapper/gather.py]: Now I replaced slice with index_select and expand_dims here.

File: pyzoo/zoo/pipeline/api/onnx/mapper/gather.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        super(GatherMapper, self).__init__(node, initializer, _all_tensors)
+
+    def _extract_model_inputs(self):
+        return [self._to_zoo_input(i) for i in self._input_list]
+
+    def _to_tensor(self):
+        data = self.model_inputs[0].zvalue
+        
+        dim = int(self.onnx_attr['axis'])
+        assert dim == 1, "Currently only dim=1 is supported"
```
Comment (zhichao-li) [inline pyzoo/zoo/pipeline/api/onnx/mapper/gather.py]: dim >=1?

File: pyzoo/zoo/pipeline/api/onnx/mapper/gather.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+        index = self.model_inputs[1].zvalue
+        assert index.shape == (1,), "Currently only one index is supported."
+
+        return zautograd.expand_dims(data.index_select(dim=dim, index=1), axis=1)
```
Comment (zhichao-li) [inline pyzoo/zoo/pipeline/api/onnx/mapper/gather.py]: index=1 ? axis = dim?

File: pyzoo/test/zoo/pipeline/onnx/test_model_loading.py [hunk 5]
```diff
@@ -600,6 +600,15 @@ def test_sigmoid(self):
         output = OnnxLoader.run_node(node, [x])
         np.testing.assert_almost_equal(output["y"], y, decimal=5)
 
+    def test_onnx_index_select(self):
+        class IndexSelect(torch.nn.Module):
+            def forward(self, x):
+                return torch.index_select(x, dim=1, index=torch.tensor([1]))
+
+        pytorch_model = IndexSelect()
```
Comment (zhichao-li) [inline pyzoo/test/zoo/pipeline/onnx/test_model_loading.py]: can you pass some parameters into IndexSelect()? i.e `IndexSelect(3, 2) and using inputShape with (3, 4, 5, 6)`

File: pyzoo/zoo/pipeline/api/onnx/mapper/gather.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        data = self.model_inputs[0].zvalue
+
+        dim = int(self.onnx_attr['axis'])
+        assert dim == 1, "Currently only dim=1 is supported."
+
+        indices = self.model_inputs[1].zvalue
+        assert indices.shape == (1,), "Currently only one index is supported."
+        index = int(indices.get_weight().max())
+
+        return zautograd.expand_dims(data.index_select(dim=dim, index=index), axis=1)
```
Comment (zhichao-li) [inline pyzoo/zoo/pipeline/api/onnx/mapper/gather.py]: why we hardcode `axis=1` here?

**Comment by cynthia-liu** [comment]
Check console output at http://10.239.47.211:18888/job/ZOO-PR-Validation/1388/ to view the results.

File: pyzoo/test/zoo/pipeline/onnx/test_model_loading.py [hunk 7]
```diff
@@ -600,6 +600,36 @@ def test_sigmoid(self):
...
+                return torch.index_select(x, dim=self.dim, index=torch.tensor(self.index))
+
+        pytorch_model = IndexSelect(3, 2)
+        input_shape_with_batch = (3, 4, 5, 6)
+        self.compare_with_pytorch(pytorch_model, input_shape_with_batch)
+
+    def test_index_select_axis0(self):
+        import pytest
+        with pytest.raises(Exception) as e_info:
+            class IndexSelect(torch.nn.Module):
```
Comment (zhichao-li) [inline pyzoo/test/zoo/pipeline/onnx/test_model_loading.py]: It would be better if we can avoid duplicating the class definition.
